src/components/app/ProfileSharing.js

```diff
@@ -137,72 +149,42 @@ class ProfileSharingCompositeButton extends PureComponent {
-        const uploadPromise = uploadBinaryProfileData(
-          gzipData,
-          uploadProgress => {
-            this.setState({ uploadProgress });
-          }
-        ).then(hash => {
-          const { onProfilePublished } = this.props;
-          onProfilePublished(hash);
-          const newShortURL =
-            this.state.fullURL === window.location.href
-              ? this.state.shortURL
-              : window.location.href;
-          this.setState({
-            state: 'public',
-            hash,
-            fullURL: window.location.href,
-            shortURL: newShortURL,
-          });
-        });
-        const shortenURLPromise = this._shortenURLAndFocusTextFieldOnCompletion();
-        Promise.race([uploadPromise, shortenURLPromise]).then(() => {
-          if (this._permalinkButton) {
-            this._permalinkButton.openPanel();
-          }
-        });
-        return Promise.all([uploadPromise, shortenURLPromise]);
-      })
-      .catch(error => {
-        this.setState({
-          state: 'error',
-          error,
-        });
-        if (this._uploadErrorButton) {
-          this._uploadErrorButton.openPanel();
-        }
-      });
+      const typedArray = await new TextEncoder().encode(jsonString);
+      const [gzipData, hash] = await Promise.all([
+        compress(typedArray.slice(0)),
+        sha1(typedArray),
+      ]);
+      const predictedUrl = url.resolve(
+        window.location.href,
+        predictUrl(uploadSuccess(hash)) // uploadSuccess is used directly, so it doesn't dispatch an action
+      );
+      const uploadPromise = uploadBinaryProfileData(gzipData);
+      const shortenUrlPromise = this._shortenUrlAndFocusTextFieldOnCompletion(
+        predictedUrl
+      );
+      await Promise.race([uploadPromise, shortenUrlPromise]);
+      if (this._permalinkButton) {
+        this._permalinkButton.openPanel();
+      }
+      await uploadPromise;
+      await shortenUrlPromise;
+    } catch (error) {
+      const { uploadError } = this.props;
+      uploadError(error);
+    }
   }
 
   render() {
-    const { state, uploadProgress, error, shortURL } = this.state;
+    const { shortUrl, dataSource, status, progress, error } = this.props;
+    const classes = classNames('profileSharingCompositeButtonContainer', {
+      currentButtonIsShareButton: status === 'none' && dataSource !== 'public',
+      currentButtonIsPermalinkButton:
+        status === 'none' && dataSource === 'public',
+      currentButtonIsUploadingButton: status === 'uploading',
+      currentButtonIsUploadErrorButton: status === 'error',
+    });
     return (
-      <div
-        className={classNames('profileSharingCompositeButtonContainer', {
-          currentButtonIsShareButton: state === 'local',
-          currentButtonIsUploadingButton: state === 'uploading',
-          currentButtonIsPermalinkButton: state === 'public',
-          currentButtonIsUploadErrorButton: state === 'error',
-        })}
-      >
```

Comment:
The `state` state was duplicating other available data, so I used the existing data directly instead.